flow/record/base.py

```diff
@@ -661,7 +691,56 @@ def DynamicDescriptor(name, fields):
     return RecordDescriptor(name, [("dynamic", field) for field in fields])
 
 
-def open_path(path, mode, clobber=True):
+def open_stream(fp: BinaryIO, mode: str) -> BinaryIO:
+    if not hasattr(fp, "peek"):
+        fp = Peekable(fp)
+
+    # We peek into the file at the maximum possible length we might need, which is the amount of bytes needed to
+    # determine whether a stream is a RECORDSTREAM or not.
+    peek_data = fp.peek(RECORDSTREAM_MAGIC_DEPTH)
+
+    # If the data stream is compressed, we wrap the file pointer in a reader that can decompress accordingly.
+    if peek_data[:2] == GZIP_MAGIC:
+        fp = gzip.GzipFile(fileobj=fp, mode=mode)
+    elif HAS_BZ2 and peek_data[:3] == BZ2_MAGIC:
+        fp = bz2.BZ2File(fp, mode=mode)
+    elif HAS_LZ4 and peek_data[:4] == LZ4_MAGIC:
+        fp = lz4.open(fp, mode=mode)
+    elif HAS_ZSTD and peek_data[:4] == ZSTD_MAGIC:
+        dctx = zstd.ZstdDecompressor()
+        fp = dctx.stream_reader(fp)
+
+    return fp
+
+
+def find_adapter_for_stream(fp: BinaryIO) -> tuple[BinaryIO, str]:
+    new_stream = Peekable(fp)
+    peek_data = new_stream.peek(RECORDSTREAM_MAGIC_DEPTH)
+    if new_stream.peeked_data is not None:
+        peek_data = new_stream.peeked_data
+    else:
+        peek_data = new_stream.peek(RECORDSTREAM_MAGIC_DEPTH)
+    if peek_data[:3] == AVRO_MAGIC and HAS_AVRO:
+        return new_stream, "avro"
+    elif RECORDSTREAM_MAGIC in peek_data:
+        return new_stream, "stream"
+    return None, None
+
+
+def open_file(path: Any, mode: str, clobber: bool = True) -> BinaryIO:
+    if isinstance(path, Path):
+        path = str(path)
+    if isinstance(path, str):
+        return open_path(path, mode, clobber)
+    elif isinstance(path, Peekable):
```

Comment:
Would this not prevent any decompression from being opened?

Reply:
Could you elaborate in what situation this would prevent decompression from being opened? As far as I understand, the existing tests already cover compression and decompression for paths, and I added tests for decompression of file-like objects, and I didn't encounter a scenario where decompression failed.

Reply:
I haven't followed all code paths, but exiting early here in case of a `Peekable` and not calling `open_stream`, I figured that if you somehow where to have a `Peekable` of a (still) compressed stream, we wouldn't do compression stream detection on it. This is probably not a realistic code path though.